katcp/test/test_client.py

```diff
@@ -310,6 +315,182 @@ def notify_connected(connected):
                         "Expected %r to not be %r" % (stream, self.client._stream))
         self.assertEqual(sockname, self.client._stream.socket.getpeername())
 
+
+class TestDeviceClientMemoryLeaks(tornado.testing.AsyncTestCase):
+
+    def setUp(self):
+        super(TestDeviceClientMemoryLeaks, self).setUp()
+        self.server = DeviceTestServer('', 0)
+        start_thread_with_cleanup(self, self.server, start_timeout=0.1)
+        self.host, self.port = self.server.bind_address
+
+    def test_no_memory_leak_managed_ioloop(self):
+        client = katcp.DeviceClient(self.host, self.port)
+        wr = weakref.ref(client)
+
+        client.start(timeout=0.1)
+        client.wait_protocol(timeout=0.1)
+        self.assertTrue(client.protocol_flags)
+        client.stop(timeout=0.1)
+        client.join(timeout=0.1)
+
+        # clear strong reference and check if object can be garbage collected
+        client = None
+        gc.collect()
+        self.assertIsNone(wr())
+
+    @tornado.testing.gen_test
+    def test_no_memory_leak_unmanaged_ioloop(self):
+        client = katcp.DeviceClient(self.host, self.port)
+        wr = weakref.ref(client)
+
+        # use test's ioloop, so client does not create its own (i.e., unmanaged)
+        client.set_ioloop(self.io_loop)
+        client.start(timeout=0.1)
+        yield client.until_protocol(timeout=0.1)
+        self.assertTrue(client.protocol_flags)
+        client.stop(timeout=0.1)
+        yield client.until_stopped(timeout=0.1)
+
+        # clear strong reference and check if object can be garbage collected
+        client = None
+        gc.collect()
+        self.assertIsNone(wr())
+
+    @tornado.testing.gen_test
+    def test_no_memory_leak_change_ioloop(self):
+        client = katcp.DeviceClient(self.host, self.port)
+        wr = weakref.ref(client)
+
+        # start and stop client with managed ioloop
+        client.start(timeout=0.1)
+        client.wait_protocol(timeout=0.1)
+        self.assertTrue(client.protocol_flags)
+        client.stop(timeout=0.1)
+        client.join(timeout=0.1)
```

Comment:
The start/wait_protocol/assert/stop/join block is copy-pasted a lot, and could go into a helper function. A similar helper could cover the unmanaged case.